// spinalcore/lib/spinalcore.node.js
/*
* Copyright 2015 SpinalCom - www.spinalcom.com
*
* This file is part of SpinalCore.
*
* Please read all of the following terms and conditions of the Free Software
* license Agreement ("Agreement") carefully.
*
* This Agreement is a legally binding contract between the Licensee (as defined
* below) and SpinalCom that sets forth the terms and conditions that govern
* your use of the Program. By installing and/or using the Program, you agree to
* abide by all the terms and conditions stated or referenced herein.
*
* If you do not agree to abide by these terms and conditions, do not
* demonstrate your acceptance and do not install or use the Program.
*
* You should have received a copy of the license along with this file. If not,
* see <http://resources.spinalcom.com/licenses.pdf>.
*/

// Generated by CoffeeScript 1.10.0
var root, url,
  indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; },
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

url = require('url');

root = typeof _root_obj === "undefined" ? global : window;

root.spinalCore = (function() {
  function spinalCore() {}

  spinalCore.connect = function(options) {
    var auth;
    if (typeof options === 'string') {
      options = url.parse(options);
    }
    if (options.path.slice(-1)[0] !== "/") {
      options.path += "/";
    }
    FileSystem._home_dir = options.path;
    FileSystem._url = options.hostname;
    FileSystem._port = options.port;
    if (options.auth !== null) {
      auth = options.auth.split(":");
      FileSystem._userid = auth[0];
      if (auth.length > 1) {
        FileSystem._password = auth[1];
      }
    } else {
      FileSystem._userid = 644;
      FileSystem._password = "";
    }
    return new FileSystem;
  };

  spinalCore.store = function(fs, model, path, callback_success, callback_error) {
    var file_name, lst;
    if (typeof callback_error === "undefined") {
      callback_error = function() {
        return console.log("Model could not be stored. You can pass a callback to handle this error.");
      };
    }
    lst = path.split("/");
    file_name = lst.pop();
    if (lst[0] === "") {
      lst.splice(0, 1);
    }
    path = lst.join("/");
    return fs.load_or_make_dir(FileSystem._home_dir + path, function(dir, err) {
      var file;
      if (err) {
        return callback_error();
      } else {
        file = dir.detect(function(x) {
          return x.name.get() === file_name;
        });
        if (file != null) {
          dir.remove(file);
        }
        dir.add_file(file_name, model, {
          model_type: "Model"
        });
        return callback_success();
      }
    });
  };

  spinalCore.load = function(fs, path, callback_success, callback_error) {
    var file_name, lst;
    if (typeof callback_error === "undefined") {
      callback_error = function() {
        return console.log("Model could not be loaded. You can pass a callback to handle this error.");
      };
    }
    lst = path.split("/");
    file_name = lst.pop();
    if (lst[0] === "") {
      lst.splice(0, 1);
    }
    path = lst.join("/");
    return fs.load_or_make_dir(FileSystem._home_dir + path, function(current_dir, err) {
      var file;
      if (err) {
        return callback_error();
      } else {
        file = current_dir.detect(function(x) {
          return x.name.get() === file_name;
        });
        if (file != null) {
          return file.load((function(_this) {
            return function(data, err) {
              if (err) {
                return callback_error();
              } else {
                return callback_success(data, err);
              }
            };
          })(this));
        } else {
          return callback_error();
        }
      }
    });
  };

  spinalCore.extend = function(child, parent) {
    var child_name, ctor, key, value;
    for (key in parent) {
      value = parent[key];
      child[key] = value;
    }
    ctor = function() {
      this.constructor = child;
    };
    ctor.prototype = parent.prototype;
    child.prototype = new ctor();
    child.__super__ = parent.prototype;
    child.super = function () {
            var args = [];
           	for (var i=1; i < arguments.length; i++)
                args[i-1] = arguments[i];
            child.__super__.constructor.apply(arguments[0], args);
        };
    root = typeof global !== "undefined" && global !== null ? global : this;
    child_name = /^function\s+([\w\$]+)\s*\(/.exec(child.toString())[1];
    return root[child_name] = child;
  };

  return spinalCore;

})();

module.exports = spinalCore;

root = typeof _root_obj === "undefined" ? global : window;

root.ModelProcessManager = (function() {
  function ModelProcessManager() {}

  ModelProcessManager._counter = 0;

  ModelProcessManager._modlist = {};

  ModelProcessManager._n_processes = {};

  ModelProcessManager._cur_mid = 0;

  ModelProcessManager._cur_process_id = 0;

  ModelProcessManager._timeout = void 0;

  ModelProcessManager._force_m = false;

  ModelProcessManager._synchro = void 0;

  ModelProcessManager.new_from_state = function(str) {
    var l, len, lst, map, mid, q, s;
    map = {};
    lst = str.split("\n");
    mid = lst.shift();
    for (q = 0, len = lst.length; q < len; q++) {
      l = lst[q];
      if (!l.length) {
        continue;
      }
      s = l.split(" ");
      map[s[0]] = {
        type: s[1],
        data: s[2],
        buff: void 0
      };
    }
    eval("var __new__ = new " + map[mid].type + ";");
    __new__._set_state(map[mid].data, map);
    return __new__;
  };

  ModelProcessManager.load = function(filename, func) {
    if (!ModelProcessManager.synchronizer) {
      ModelProcessManager._synchro = new Synchronizer;
    }
    return ModelProcessManager._synchro.load(filename, func);
  };

  ModelProcessManager.conv = function(v) {
    if (v instanceof Model) {
      return v;
    }
    if (v instanceof Array) {
      return new Lst(v);
    }
    if (typeof v === "string") {
      return new Str(v);
    }
    if (typeof v === "number") {
      return new Val(v);
    }
    if (typeof v === "boolean") {
      return new Bool(v);
    }
    if (v instanceof Object) {
      return new Model(v);
    }
    return new Obj(v);
  };

  ModelProcessManager.get_object_class = function(obj) {
    var arr;
    if (obj && obj.constructor && obj.constructor.toString) {
      arr = obj.constructor.toString().match(/function\s*(\w+)/);
      if (arr && arr.length === 2) {
        return arr[1];
      }
    }
  };

  ModelProcessManager._get_attribute_names = function(m) {
    var key, results, val;
    if (m instanceof Model) {
      return m._attribute_names;
    } else {
      results = [];
      for (key in m) {
        val = m[key];
        results.push(key);
      }
      return results;
    }
  };

  ModelProcessManager._new_model_from_state = function(mid, map) {
    var info;
    info = map[mid];
    eval("info.buff = new " + info.type + ";");
    info.buff._set_state(info.data, map);
    return info.buff;
  };

  ModelProcessManager._need_sync_processes = function() {
    if (ModelProcessManager._timeout == null) {
      return ModelProcessManager._timeout = setTimeout(ModelProcessManager._sync_processes, 1);
    }
  };

  ModelProcessManager._sync_processes = function() {
    var id, len, model, process, processes, q, ref, ref1, ref2;
    processes = {};
    ref = ModelProcessManager._modlist;
    for (id in ref) {
      model = ref[id];
      ref1 = model._processes;
      for (q = 0, len = ref1.length; q < len; q++) {
        process = ref1[q];
        processes[process.process_id] = {
          value: process,
          force: false
        };
      }
    }
    ref2 = ModelProcessManager._n_processes;
    for (id in ref2) {
      process = ref2[id];
      processes[id] = {
        value: process,
        force: true
      };
    }
    ModelProcessManager._timeout = void 0;
    ModelProcessManager._modlist = {};
    ModelProcessManager._n_processes = {};
    ModelProcessManager._counter += 2;
    for (id in processes) {
      process = processes[id];
      ModelProcessManager._force_m = process.force;
      process.value.onchange();
    }
    return ModelProcessManager._force_m = false;
  };

  return ModelProcessManager;

})();

root = typeof _root_obj === "undefined" ? global : window;

root.Model = (function() {
  function Model(attr) {
    this._attribute_names = [];
    this.model_id = ModelProcessManager._cur_mid;
    ModelProcessManager._cur_mid += 1;
    this._processes = [];
    this._parents = [];
    this._date_last_modification = ModelProcessManager._counter + 2;
    if (attr != null) {
      this._set(attr);
    }
  }

  Model.prototype.destructor = function() {};

  Model.prototype.has_been_modified = function() {
    return this._date_last_modification > ModelProcessManager._counter - 2 || ModelProcessManager._force_m;
  };

  Model.prototype.has_been_directly_modified = function() {
    return this._date_last_modification > ModelProcessManager._counter - 1 || ModelProcessManager._force_m;
  };

  Model.prototype.bind = function(f, onchange_construction) {
    if (onchange_construction == null) {
      onchange_construction = true;
    }
    if (f instanceof Process) {
      this._processes.push(f);
      f._models.push(this);
      if (onchange_construction) {
        ModelProcessManager._n_processes[f.process_id] = f;
        return ModelProcessManager._need_sync_processes();
      }
    } else {
      return new BindProcess(this, onchange_construction, f);
    }
  };

  Model.prototype.unbind = function(f) {
    var len, q, ref, results, v;
    if (f instanceof Process) {
      this._processes.splice(this._processes.indexOf(f), 1);
      return f._models.splice(f._models.indexOf(this), 1);
    } else {
      ref = this._processes;
      results = [];
      for (q = 0, len = ref.length; q < len; q++) {
        v = ref[q];
        if (v instanceof BindProcess && v.f === f) {
          results.push(this.unbind(v));
        }
      }
      return results;
    }
  };

  Model.prototype.get = function() {
    var len, name, q, ref, res;
    res = {};
    ref = this._attribute_names;
    for (q = 0, len = ref.length; q < len; q++) {
      name = ref[q];
      res[name] = this[name].get();
    }
    return res;
  };

  Model.prototype.set = function(value) {
    if (this._set(value)) {
      this._signal_change();
      return true;
    }
    return false;
  };

  Model.prototype.set_state = function(str) {
    var l, len, lst, map, mid, q, s;
    map = {};
    lst = str.split("\n");
    mid = lst.shift();
    for (q = 0, len = lst.length; q < len; q++) {
      l = lst[q];
      if (!l.length) {
        continue;
      }
      s = l.split(" ");
      map[s[0]] = {
        type: s[1],
        data: s[2],
        buff: void 0
      };
    }
    map[mid].buff = this;
    return this._set_state(map[mid].data, map);
  };

  Model.prototype.get_state = function(date) {
    var fmm, id, obj, res;
    if (date == null) {
      date = -1;
    }
    fmm = {};
    this._get_flat_model_map(fmm, date);
    res = this.model_id.toString();
    if (this._date_last_modification > date) {
      for (id in fmm) {
        obj = fmm[id];
        res += "\n" + obj.model_id + " " + ModelProcessManager.get_object_class(obj) + " " + obj._get_state();
      }
    }
    return res;
  };

  Model.prototype.add_attr = function(n, p, signal_change) {
    var key, results, val;
    if (signal_change == null) {
      signal_change = true;
    }
    if (p != null) {
      if (typeof p === "function") {
        return this[n] = p;
      } else {
        if (this[n] != null) {
          console.error("attribute " + n + " already exists in " + (ModelProcessManager.get_object_class(this)));
        }
        p = ModelProcessManager.conv(p);
        if (indexOf.call(p._parents, this) < 0) {
          p._parents.push(this);
        }
        this._attribute_names.push(n);
        this[n] = p;
        if (signal_change) {
          return this._signal_change();
        }
      }
    } else {
      results = [];
      for (key in n) {
        val = n[key];
        if (val != null) {
          results.push(this.add_attr(key, val, signal_change));
        }
      }
      return results;
    }
  };

  Model.prototype.rem_attr = function(name, signal_change) {
    var c, i;
    if (signal_change == null) {
      signal_change = true;
    }
    c = this[name];
    if (c) {
      i = c._parents.indexOf(this);
      if (i >= 0) {
        c._parents.splice(i, 1);
        if (c._parents.length === 0) {
          c.destructor();
        }
      }
      delete this[name];
      i = this._attribute_names.indexOf(name);
      if (i >= 0) {
        this._attribute_names.splice(i, 1);
      }
      if (signal_change) {
        return this._signal_change();
      }
    }
  };

  Model.prototype.mod_attr = function(n, p) {
    if (this[n] !== p) {
      this.rem_attr(n);
      return this.add_attr(n, p);
    }
  };

  Model.prototype.set_attr = function(o) {
    var k, len, q, r, results, to_rem, v;
    for (k in o) {
      v = o[k];
      this.mod_attr(k, v);
    }
    to_rem = (function() {
      var len, q, ref, results;
      ref = this._attribute_names;
      results = [];
      for (q = 0, len = ref.length; q < len; q++) {
        k = ref[q];
        if (o[k] == null) {
          results.push(k);
        }
      }
      return results;
    }).call(this);
    results = [];
    for (q = 0, len = to_rem.length; q < len; q++) {
      r = to_rem[q];
      results.push(this.rem_attr(r));
    }
    return results;
  };

  Model.prototype.size = function(for_display) {
    if (for_display == null) {
      for_display = false;
    }
    return [];
  };

  Model.prototype.dim = function(for_display) {
    if (for_display == null) {
      for_display = false;
    }
    return this.size(for_display).length;
  };

  Model.prototype.equals = function(m) {
    var key, len, len1, q, ref, ref1, t, u, val;
    if (this === m) {
      return true;
    }
    if (m._attribute_names != null) {
      u = {};
      ref = m._attribute_names;
      for (q = 0, len = ref.length; q < len; q++) {
        key = ref[q];
        val = m[key];
        if (this[key] == null) {
          return false;
        }
        if (!this[key].equals(val)) {
          return false;
        }
        u[key] = true;
      }
      ref1 = this._attribute_names;
      for (t = 0, len1 = ref1.length; t < len1; t++) {
        key = ref1[t];
        if (u[key] == null) {
          return false;
        }
      }
    }
    return false;
  };

  Model.prototype.get_parents_that_check = function(func_to_check) {
    var res, visited;
    res = [];
    visited = {};
    this._get_parents_that_check_rec(res, visited, func_to_check);
    return res;
  };

  Model.prototype.deep_copy = function() {
    var key, len, o, q, ref;
    o = {};
    ref = this._attribute_names;
    for (q = 0, len = ref.length; q < len; q++) {
      key = ref[q];
      o[key] = this[key].deep_copy();
    }
    eval("var __new__ = new " + (ModelProcessManager.get_object_class(this)) + ";");
    __new__.set_attr(o);
    return __new__;
  };

  Model.prototype.real_change = function() {
    var a, len, q, ref;
    if (this.has_been_directly_modified() && !this._attribute_names.length) {
      return true;
    }
    ref = this._attribute_names;
    for (q = 0, len = ref.length; q < len; q++) {
      a = ref[q];
      if (typeof this.cosmetic_attribute === "function" ? this.cosmetic_attribute(a) : void 0) {
        continue;
      }
      if (this[a].real_change()) {
        return true;
      }
    }
    return false;
  };

  Model.prototype.cosmetic_attribute = function(name) {
    return false;
  };

  Model.prototype._get_state = function() {
    var name, str;
    str = (function() {
      var len, q, ref, results;
      ref = this._attribute_names;
      results = [];
      for (q = 0, len = ref.length; q < len; q++) {
        name = ref[q];
        results.push(name + ":" + this[name].model_id);
      }
      return results;
    }).call(this);
    return str.join(",");
  };

  Model.prototype._get_fs_data = function(out) {
    var name, obj, str;
    FileSystem.set_server_id_if_necessary(out, this);
    str = (function() {
      var len, q, ref, results;
      ref = this._attribute_names;
      results = [];
      for (q = 0, len = ref.length; q < len; q++) {
        name = ref[q];
        obj = this[name];
        FileSystem.set_server_id_if_necessary(out, obj);
        results.push(name + ":" + obj._server_id);
      }
      return results;
    }).call(this);
    return out.mod += "C " + this._server_id + " " + (str.join(",")) + " ";
  };

  Model.prototype._set = function(value) {
    var change, key, len, len1, q, ref, ref1, t, used, val;
    change = false;
    used = {};
    ref = ModelProcessManager._get_attribute_names(value);
    for (q = 0, len = ref.length; q < len; q++) {
      key = ref[q];
      used[key] = true;
    }
    ref1 = (function() {
      var len1, ref1, results, w;
      ref1 = this._attribute_names;
      results = [];
      for (w = 0, len1 = ref1.length; w < len1; w++) {
        key = ref1[w];
        if (!used[key]) {
          results.push(key);
        }
      }
      return results;
    }).call(this);
    for (t = 0, len1 = ref1.length; t < len1; t++) {
      key = ref1[t];
      change = true;
      this.rem_attr(key, false);
    }
    for (key in value) {
      val = value[key];
      if (val != null) {
        if (this[key] != null) {
          if (this[key].constructor === val.constructor) {
            change |= this[key].set(val);
          } else {
            change = true;
            this.mod_attr(key, val, false);
          }
        } else {
          this.add_attr(key, val, false);
        }
      }
    }
    return change;
  };

  Model.prototype._signal_change = function(change_level) {
    var len, p, q, ref;
    if (change_level == null) {
      change_level = 2;
    }
    if (change_level === 2 && (this._server_id != null)) {
      FileSystem.signal_change(this);
    }
    ModelProcessManager._modlist[this.model_id] = this;
    if (this._date_last_modification <= ModelProcessManager._counter) {
      this._date_last_modification = ModelProcessManager._counter + change_level;
      ref = this._parents;
      for (q = 0, len = ref.length; q < len; q++) {
        p = ref[q];
        p._signal_change(1);
      }
    }
    return ModelProcessManager._need_sync_processes();
  };

  Model.prototype._set_state = function(str, map) {
    var attr, inr, k_id, len, len1, q, ref, ref1, results, spl, t, u;
    u = {};
    if (str.length) {
      ref = str.split(",");
      for (q = 0, len = ref.length; q < len; q++) {
        spl = ref[q];
        inr = spl.split(":");
        attr = inr[0];
        k_id = inr[1];
        u[attr] = true;
        if (map[k_id].buff != null) {
          if (this[attr] == null) {
            this.add_attr(attr, map[k_id].buff);
          } else if (map[k_id].buff !== this[attr]) {
            this.mod_attr(attr, map[k_id].buff);
          }
        } else if (this[attr] == null) {
          this.add_attr(attr, ModelProcessManager._new_model_from_state(k_id, map));
        } else if (!this[attr]._set_state_if_same_type(k_id, map)) {
          this.mod_attr(attr, ModelProcessManager._new_model_from_state(k_id, map));
        }
      }
    }
    ref1 = this._attribute_names;
    results = [];
    for (t = 0, len1 = ref1.length; t < len1; t++) {
      attr = ref1[t];
      if (!u[attr]) {
        results.push(this.rem_attr(attr));
      } else {
        results.push(void 0);
      }
    }
    return results;
  };

  Model.prototype._get_parents_that_check_rec = function(res, visited, func_to_check) {
    var len, p, q, ref, results;
    if (visited[this.model_id] == null) {
      visited[this.model_id] = true;
      if (func_to_check(this)) {
        return res.push(this);
      } else {
        ref = this._parents;
        results = [];
        for (q = 0, len = ref.length; q < len; q++) {
          p = ref[q];
          results.push(p._get_parents_that_check_rec(res, visited, func_to_check));
        }
        return results;
      }
    }
  };

  Model.prototype._set_state_if_same_type = function(mid, map) {
    var dat;
    dat = map[mid];
    if (ModelProcessManager.get_object_class(this) === dat.type) {
      dat.buff = this;
      this._set_state(dat.data, map);
      return true;
    }
    return false;
  };

  Model.prototype._get_flat_model_map = function(map, date) {
    var len, name, obj, q, ref, results;
    map[this.model_id] = this;
    ref = this._attribute_names;
    results = [];
    for (q = 0, len = ref.length; q < len; q++) {
      name = ref[q];
      obj = this[name];
      if (map[obj.model_id] == null) {
        if (obj._date_last_modification > date) {
          results.push(obj._get_flat_model_map(map, date));
        } else {
          results.push(void 0);
        }
      } else {
        results.push(void 0);
      }
    }
    return results;
  };

  return Model;

})();

root = typeof _root_obj === "undefined" ? global : window;

root.Obj = (function(superClass) {
  extend(Obj, superClass);

  function Obj(data) {
    Obj.__super__.constructor.call(this);
    if (data != null) {
      this._set(data);
    }
  }

  Obj.prototype.toString = function() {
    var ref;
    return (ref = this._data) != null ? ref.toString() : void 0;
  };

  Obj.prototype.equals = function(obj) {
    if (obj instanceof Obj) {
      return this._data === obj._data;
    }
    return this._data === obj;
  };

  Obj.prototype.get = function() {
    return this._data;
  };

  Obj.prototype._get_fs_data = function(out) {
    FileSystem.set_server_id_if_necessary(out, this);
    return out.mod += "C " + this._server_id + " " + (this.toString()) + " ";
  };

  Obj.prototype._set = function(value) {
    if (this._data !== value) {
      this._data = value;
      return true;
    }
    return false;
  };

  Obj.prototype._get_state = function() {
    return this._data;
  };

  Obj.prototype._set_state = function(str, map) {
    return this.set(str);
  };

  return Obj;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.Choice = (function(superClass) {
  extend(Choice, superClass);

  function Choice(data, initial_list) {
    if (initial_list == null) {
      initial_list = [];
    }
    Choice.__super__.constructor.call(this);
    this.add_attr({
      num: 0,
      lst: initial_list
    });
    if (data != null) {
      this.num.set(data);
    }
  }

  Choice.prototype.filter = function(obj) {
    return true;
  };

  Choice.prototype.item = function() {
    return this._nlst()[this.num.get()];
  };

  Choice.prototype.get = function() {
    var ref;
    return (ref = this.item()) != null ? ref.get() : void 0;
  };

  Choice.prototype.toString = function() {
    var ref;
    return (ref = this.item()) != null ? ref.toString() : void 0;
  };

  Choice.prototype.equals = function(a) {
    if (a instanceof Choice) {
      return Choice.__super__.equals.call(this, a);
    } else {
      return this._nlst()[this.num.get()].equals(a);
    }
  };

  Choice.prototype._set = function(value) {
    var i, j, len, q, ref;
    ref = this._nlst();
    for (j = q = 0, len = ref.length; q < len; j = ++q) {
      i = ref[j];
      if (i.equals(value)) {
        return this.num.set(j);
      }
    }
    return this.num.set(value);
  };

  Choice.prototype._nlst = function() {
    var l, len, q, ref, results;
    ref = this.lst;
    results = [];
    for (q = 0, len = ref.length; q < len; q++) {
      l = ref[q];
      if (this.filter(l)) {
        results.push(l);
      }
    }
    return results;
  };

  return Choice;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.Bool = (function(superClass) {
  extend(Bool, superClass);

  function Bool(data) {
    Bool.__super__.constructor.call(this);
    this._data = false;
    if (data != null) {
      this._set(data);
    }
  }

  Bool.prototype.toggle = function() {
    return this.set(!this._data);
  };

  Bool.prototype.toBoolean = function() {
    return this._data;
  };

  Bool.prototype.deep_copy = function() {
    return new Bool(this._data);
  };

  Bool.prototype._set = function(value) {
    var n;
    if (n instanceof Model) {
      n = value.toBoolean();
    } else if (value === "false") {
      n = false;
    } else if (value === "true") {
      n = true;
    } else {
      n = Boolean(value);
    }
    if (this._data !== n) {
      this._data = n;
      return true;
    }
    return false;
  };

  Bool.prototype._get_fs_data = function(out) {
    FileSystem.set_server_id_if_necessary(out, this);
    return out.mod += "C " + this._server_id + " " + (1 * Boolean(this._data)) + " ";
  };

  return Bool;

})(Obj);

root = typeof _root_obj === "undefined" ? global : window;

root.ConstOrNotModel = (function(superClass) {
  extend(ConstOrNotModel, superClass);

  function ConstOrNotModel(bool, model, check_disabled) {
    if (check_disabled == null) {
      check_disabled = true;
    }
    ConstOrNotModel.__super__.constructor.call(this);
    this.add_attr({
      bool: bool,
      model: model,
      check_disabled: check_disabled
    });
  }

  ConstOrNotModel.prototype.get = function() {
    var ref;
    return (ref = this.model) != null ? ref.get() : void 0;
  };

  ConstOrNotModel.prototype.set = function(value) {
    var ref;
    return (ref = this.model) != null ? ref.set(value) : void 0;
  };

  ConstOrNotModel.prototype.toString = function() {
    var ref;
    return (ref = this.model) != null ? ref.toString() : void 0;
  };

  return ConstOrNotModel;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.ConstrainedVal = (function(superClass) {
  extend(ConstrainedVal, superClass);

  function ConstrainedVal(value, params) {
    if (params == null) {
      params = {};
    }
    ConstrainedVal.__super__.constructor.call(this);
    this.add_attr({
      val: value || 0,
      _min: params.min != null ? params.min : 0,
      _max: params.max != null ? params.max : 100
    });
    this.add_attr({
      _div: params.div != null ? params.div : this._max - this._min
    });
  }

  ConstrainedVal.prototype.get = function() {
    return this.val.get();
  };

  ConstrainedVal.prototype.ratio = function() {
    return (this.val.get() - this._min.get()) / this.delta();
  };

  ConstrainedVal.prototype.delta = function() {
    return this._max.get() - this._min.get();
  };

  ConstrainedVal.prototype.set_params = function(params) {
    this._min.set(params.min != null ? params.min : 0);
    this._max.set(params.max != null ? params.max : 100);
    return this._div.set(params.div != null ? params.div : this._max - this._min);
  };

  ConstrainedVal.prototype._set = function(value) {
    var res;
    if (value instanceof ConstrainedVal) {
      return this.val._set(value.get());
    }
    res = this.val.set(value);
    this._check_val();
    return res;
  };

  ConstrainedVal.prototype._check_val = function() {
    var d, m, n, r, s, v;
    v = this.val.get();
    m = this._min.get();
    n = this._max.get();
    d = this._div.get();
    if (v < m) {
      this.val.set(m);
    }
    if (v > n) {
      this.val.set(n);
    }
    if (d) {
      s = (n - m) / d;
      r = m + Math.round((this.val.get() - m) / s) * s;
      return this.val.set(r);
    }
  };

  return ConstrainedVal;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.Lst = (function(superClass) {
  extend(Lst, superClass);

  function Lst(data) {
    var d, i, q, ref, s;
    Lst.__super__.constructor.call(this);
    this.length = 0;
    s = this.static_length();
    if (s >= 0) {
      d = this.default_value();
      for (i = q = 0, ref = s; 0 <= ref ? q < ref : q > ref; i = 0 <= ref ? ++q : --q) {
        this.push(d, true);
      }
    }
    if (data != null) {
      this._set(data);
    }
  }

  Lst.prototype.static_length = function() {
    return -1;
  };

  Lst.prototype.default_value = function() {
    return 0;
  };

  Lst.prototype.base_type = function() {
    return void 0;
  };

  Lst.prototype.get = function() {
    var i, len, q, results;
    results = [];
    for (q = 0, len = this.length; q < len; q++) {
      i = this[q];
      results.push(i.get());
    }
    return results;
  };

  Lst.prototype.size = function() {
    return [length];
  };

  Lst.prototype.toString = function() {
    var l, x;
    if (this.length) {
      l = (function() {
        var len, q, results;
        results = [];
        for (q = 0, len = this.length; q < len; q++) {
          x = this[q];
          results.push(x.toString());
        }
        return results;
      }).call(this);
      return l.join();
    } else {
      return "";
    }
  };

  Lst.prototype.equals = function(lst) {
    var i, q, ref;
    if (this.length !== lst.length) {
      return false;
    }
    for (i = q = 0, ref = this.length; 0 <= ref ? q < ref : q > ref; i = 0 <= ref ? ++q : --q) {
      if (!this[i].equals(lst[i])) {
        return false;
      }
    }
    return true;
  };

  Lst.prototype.push = function(value, force) {
    var b;
    if (force == null) {
      force = false;
    }
    if (this._static_size_check(force)) {
      return;
    }
    b = this.base_type();
    if (b != null) {
      if (!(value instanceof b)) {
        value = new b(value);
      }
    } else {
      value = ModelProcessManager.conv(value);
    }
    if (indexOf.call(value._parents, this) < 0) {
      value._parents.push(this);
    }
    this[this.length] = value;
    this.length += 1;
    return this._signal_change();
  };

  Lst.prototype.pop = function() {
    var old;
    if (this._static_size_check(false)) {
      return;
    }
    if (this.length <= 0) {
      return;
    }
    this.length -= 1;
    old = this[this.length];
    this.rem_attr(this.length);
    return old;
  };

  Lst.prototype.clear = function() {
    var results;
    results = [];
    while (this.length) {
      results.push(this.pop());
    }
    return results;
  };

  Lst.prototype.unshift = function(element) {
    var b, i, q, ref;
    if (this._static_size_check(false)) {
      return;
    }
    b = this.base_type();
    if (b != null) {
      if (!(element instanceof b)) {
        element = new b(element);
      }
    } else {
      element = ModelProcessManager.conv(element);
    }
    if (indexOf.call(element._parents, this) < 0) {
      element._parents.push(this);
    }
    if (this.length) {
      for (i = q = ref = this.length - 1; ref <= 0 ? q <= 0 : q >= 0; i = ref <= 0 ? ++q : --q) {
        this[i + 1] = this[i];
      }
    }
    this[0] = element;
    this.length += 1;
    this._signal_change();
    return this.length;
  };

  Lst.prototype.shift = function() {
    var r;
    r = this[0];
    this.splice(0, 1);
    return r;
  };

  Lst.prototype.remove = function(item) {
    var i;
    i = this.indexOf(item);
    if (i >= 0) {
      return this.splice(i, 1);
    }
  };

  Lst.prototype.remove_ref = function(item) {
    var i;
    i = this.indexOf_ref(item);
    if (i >= 0) {
      return this.splice(i, 1);
    }
  };

  Lst.prototype.filter = function(f) {
    var i, len, q, results;
    results = [];
    for (q = 0, len = this.length; q < len; q++) {
      i = this[q];
      if (f(i)) {
        results.push(i);
      }
    }
    return results;
  };

  Lst.prototype.detect = function(f) {
    var i, len, q;
    for (q = 0, len = this.length; q < len; q++) {
      i = this[q];
      if (f(i)) {
        return i;
      }
    }
    return void 0;
  };

  Lst.prototype.sorted = function(fun_sort) {
    var it, len, new_array, q;
    new_array = new Array;
    for (q = 0, len = this.length; q < len; q++) {
      it = this[q];
      new_array.push(it);
    }
    new_array.sort(fun_sort);
    return new_array;
  };

  Lst.prototype.has = function(f) {
    var i, len, q;
    for (q = 0, len = this.length; q < len; q++) {
      i = this[q];
      if (f(i)) {
        return true;
      }
    }
    return false;
  };

  Lst.prototype.indexOf = function(v) {
    var i, q, ref;
    for (i = q = 0, ref = this.length; 0 <= ref ? q < ref : q > ref; i = 0 <= ref ? ++q : --q) {
      if (this[i].equals(v)) {
        return i;
      }
    }
    return -1;
  };

  Lst.prototype.indexOf_ref = function(v) {
    var i, q, ref;
    for (i = q = 0, ref = this.length; 0 <= ref ? q < ref : q > ref; i = 0 <= ref ? ++q : --q) {
      if (this[i] === v) {
        return i;
      }
    }
    return -1;
  };

  Lst.prototype.contains = function(v) {
    return this.indexOf(v) >= 0;
  };

  Lst.prototype.contains_ref = function(v) {
    return this.indexOf_ref(v) >= 0;
  };

  Lst.prototype.toggle = function(v) {
    var i;
    i = this.indexOf(v);
    if (i >= 0) {
      this.splice(i);
      return false;
    } else {
      this.push(v);
      return true;
    }
  };

  Lst.prototype.toggle_ref = function(v) {
    var i;
    i = this.indexOf_ref(v);
    if (i >= 0) {
      this.splice(i);
      return false;
    } else {
      this.push(v);
      return true;
    }
  };

  Lst.prototype.slice = function(begin, end) {
    var i, q, ref, ref1, tab;
    if (end == null) {
      end = this.length;
    }
    if (begin < 0) {
      begin = 0;
    }
    if (end > this.length) {
      end = this.length;
    }
    tab = new Lst;
    for (i = q = ref = begin, ref1 = end; ref <= ref1 ? q < ref1 : q > ref1; i = ref <= ref1 ? ++q : --q) {
      tab.push(this[i].get());
    }
    return tab;
  };

  Lst.prototype.concat = function(new_tab, force) {
    var el, len, q;
    if (force == null) {
      force = false;
    }
    if (this._static_size_check(force)) {
      return;
    }
    if (new_tab.length) {
      for (q = 0, len = new_tab.length; q < len; q++) {
        el = new_tab[q];
        this.push(el);
      }
      return this;
    }
  };

  Lst.prototype.splice = function(index, n) {
    var i, q, ref, ref1, ref2, ref3, ref4, ref5, t, w;
    if (n == null) {
      n = 1;
    }
    if (this._static_size_check(false)) {
      return;
    }
    for (i = q = ref = index, ref1 = Math.min(index + n, this.length); ref <= ref1 ? q < ref1 : q > ref1; i = ref <= ref1 ? ++q : --q) {
      this.rem_attr(i);
    }
    for (i = t = ref2 = index, ref3 = this.length - n; ref2 <= ref3 ? t < ref3 : t > ref3; i = ref2 <= ref3 ? ++t : --t) {
      this[i] = this[i + n];
    }
    for (i = w = ref4 = this.length - n, ref5 = this.length; ref4 <= ref5 ? w < ref5 : w > ref5; i = ref4 <= ref5 ? ++w : --w) {
      delete this[i];
    }
    this.length -= n;
    return this._signal_change();
  };

  Lst.prototype.insert = function(index, list) {
    var i, l, len, len1, o, q, results, t;
    if (list.length) {
      l = Math.max(this.length - index, 0);
      o = (function() {
        var q, ref, results;
        results = [];
        for (i = q = 0, ref = l; 0 <= ref ? q < ref : q > ref; i = 0 <= ref ? ++q : --q) {
          results.push(this.pop());
        }
        return results;
      }).call(this);
      o.reverse();
      for (q = 0, len = list.length; q < len; q++) {
        l = list[q];
        this.push(l);
      }
      results = [];
      for (t = 0, len1 = o.length; t < len1; t++) {
        l = o[t];
        results.push(this.push(l));
      }
      return results;
    }
  };

  Lst.prototype.set_or_push = function(index, val) {
    if (index < this.length) {
      return this.mod_attr(index, val);
    } else if (index === this.length) {
      return this.push(val);
    }
  };

  Lst.prototype.trim = function(size) {
    var results;
    results = [];
    while (this.length > size) {
      results.push(this.pop());
    }
    return results;
  };

  Lst.prototype.join = function(sep) {
    return this.get().join(sep);
  };

  Lst.prototype.deep_copy = function() {
    var i, q, ref, res;
    res = new Lst;
    for (i = q = 0, ref = this.length; 0 <= ref ? q < ref : q > ref; i = 0 <= ref ? ++q : --q) {
      res.push(this[i].deep_copy());
    }
    return res;
  };

  Lst.prototype.back = function() {
    return this[this.length - 1];
  };

  Lst.prototype.real_change = function() {
    var a, len, q;
    if (this.has_been_directly_modified()) {
      return true;
    }
    for (q = 0, len = this.length; q < len; q++) {
      a = this[q];
      if (a.real_change()) {
        return true;
      }
    }
    return false;
  };

  Lst.prototype._set = function(value) {
    var change, p, q, ref, s;
    change = this.length !== value.length;
    s = this.static_length();
    if (s >= 0 && change) {
      console.error("resizing a static array (type " + (ModelProcessManager.get_object_class(this)) + ") is forbidden");
    }
    for (p = q = 0, ref = value.length; 0 <= ref ? q < ref : q > ref; p = 0 <= ref ? ++q : --q) {
      if (p < this.length) {
        change |= this[p].set(value[p]);
      } else if (s < 0) {
        this.push(value[p]);
      }
    }
    if (s < 0) {
      while (this.length > value.length) {
        this.pop();
      }
      this.length = value.length;
    }
    return change;
  };

  Lst.prototype._get_flat_model_map = function(map, date) {
    var len, obj, q, results;
    map[this.model_id] = this;
    results = [];
    for (q = 0, len = this.length; q < len; q++) {
      obj = this[q];
      if (map[obj.model_id] == null) {
        if (obj._date_last_modification > date) {
          results.push(obj._get_flat_model_map(map, date));
        } else {
          results.push(void 0);
        }
      } else {
        results.push(void 0);
      }
    }
    return results;
  };

  Lst.prototype._get_fs_data = function(out) {
    var obj, str;
    FileSystem.set_server_id_if_necessary(out, this);
    str = (function() {
      var len, q, results;
      results = [];
      for (q = 0, len = this.length; q < len; q++) {
        obj = this[q];
        FileSystem.set_server_id_if_necessary(out, obj);
        results.push(obj._server_id);
      }
      return results;
    }).call(this);
    return out.mod += "C " + this._server_id + " " + (str.join(",")) + " ";
  };

  Lst.prototype._get_state = function() {
    var obj, str;
    str = (function() {
      var len, q, results;
      results = [];
      for (q = 0, len = this.length; q < len; q++) {
        obj = this[q];
        results.push(obj.model_id);
      }
      return results;
    }).call(this);
    return str.join(",");
  };

  Lst.prototype._set_state = function(str, map) {
    var attr, k_id, l_id, q, ref, ref1, ref2, results, t;
    l_id = str.split(",").filter(function(x) {
      return x.length;
    });
    while (this.length > l_id.length) {
      this.pop();
    }
    for (attr = q = 0, ref = this.length; 0 <= ref ? q < ref : q > ref; attr = 0 <= ref ? ++q : --q) {
      k_id = l_id[attr];
      if (map[k_id].buff != null) {
        if (map[k_id].buff !== this[attr]) {
          this.mod_attr(attr, map[k_id].buff);
        }
      } else if (!this[attr]._set_state_if_same_type(k_id, map)) {
        this.mod_attr(attr, ModelProcessManager._new_model_from_state(k_id, map));
      }
    }
    results = [];
    for (attr = t = ref1 = this.length, ref2 = l_id.length; ref1 <= ref2 ? t < ref2 : t > ref2; attr = ref1 <= ref2 ? ++t : --t) {
      k_id = l_id[attr];
      if (map[k_id].buff != null) {
        results.push(this.push(map[k_id].buff));
      } else {
        results.push(this.push(ModelProcessManager._new_model_from_state(k_id, map)));
      }
    }
    return results;
  };

  Lst.prototype._static_size_check = function(force) {
    if (this.static_length() >= 0 && !force) {
      console.error("resizing a static array (type " + (ModelProcessManager.get_object_class(this)) + ") is forbidden");
      return true;
    }
    return false;
  };

  return Lst;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.Val = (function(superClass) {
  extend(Val, superClass);

  function Val(data) {
    Val.__super__.constructor.call(this);
    this._data = 0;
    if (data != null) {
      this._set(data);
    }
  }

  Val.prototype.toggle = function() {
    return this.set(!this._data);
  };

  Val.prototype.toBoolean = function() {
    return Boolean(this._data);
  };

  Val.prototype.deep_copy = function() {
    return new Val(this._data);
  };

  Val.prototype.add = function(v) {
    if (v) {
      this._data += v;
      return this._signal_change();
    }
  };

  Val.prototype._set = function(value) {
    var n;
    if (typeof value === "string") {
      if (value.slice(0, 2) === "0x") {
        n = parseInt(value, 16);
      } else {
        n = parseFloat(value);
        if (isNaN(n)) {
          n = parseInt(value);
        }
        if (isNaN(n)) {
          console.log("Don't know how to transform " + value + " to a Val");
        }
      }
    } else if (typeof value === "boolean") {
      n = 1 * value;
    } else if (value instanceof Val) {
      n = value._data;
    } else {
      n = value;
    }
    if (this._data !== n) {
      this._data = n;
      return true;
    }
    return false;
  };

  return Val;

})(Obj);

root = typeof _root_obj === "undefined" ? global : window;

root.Vec = (function(superClass) {
  extend(Vec, superClass);

  function Vec(data) {
    Vec.__super__.constructor.call(this, data);
  }

  Vec.prototype.base_type = function() {
    return Val;
  };

  Vec.prototype._underlying_fs_type = function() {
    return "Lst";
  };

  return Vec;

})(Lst);

root = typeof _root_obj === "undefined" ? global : window;

root.Str = (function(superClass) {
  extend(Str, superClass);

  function Str(data) {
    Str.__super__.constructor.call(this);
    this._data = "";
    this.length = 0;
    if (data != null) {
      this._set(data);
    }
  }

  Str.prototype.toggle = function(str, space) {
    var i, l;
    if (space == null) {
      space = " ";
    }
    l = this._data.split(space);
    i = l.indexOf(str);
    if (i < 0) {
      l.push(str);
    } else {
      l.splice(i, 1);
    }
    return this.set(l.join(" "));
  };

  Str.prototype.contains = function(str) {
    return this._data.indexOf(str) >= 0;
  };

  Str.prototype.equals = function(str) {
    return this._data === str.toString();
  };

  Str.prototype.ends_with = function(str) {
    var l;
    l = this._data.match(str + "$");
    return (l != null ? l.length : void 0) && l[0] === str;
  };

  Str.prototype.deep_copy = function() {
    return new Str(this._data + "");
  };

  Str.prototype._get_fs_data = function(out) {
    FileSystem.set_server_id_if_necessary(out, this);
    return out.mod += "C " + this._server_id + " " + (encodeURI(this._data)) + " ";
  };

  Str.prototype._set = function(value) {
    var n;
    if (value == null) {
      return this._set("");
    }
    n = value.toString();
    if (this._data !== n) {
      this._data = n;
      this.length = this._data.length;
      return true;
    }
    return false;
  };

  Str.prototype._get_state = function() {
    return encodeURI(this._data);
  };

  Str.prototype._set_state = function(str, map) {
    return this.set(decodeURIComponent(str));
  };

  return Str;

})(Obj);

root = typeof _root_obj === "undefined" ? global : window;

root.TypedArray = (function(superClass) {
  extend(TypedArray, superClass);

  function TypedArray(size, data) {
    var B;
    TypedArray.__super__.constructor.call(this);
    if (size == null) {
      size = [];
    }
    if (!size.length) {
      size = [size];
    }
    this._size = size;
    if (data == null) {
      B = this.base_type();
      data = new B(this.nb_items());
    }
    this._data = data;
  }

  TypedArray.prototype.base_type = function() {};

  TypedArray.prototype.dim = function() {
    return this._size.length;
  };

  TypedArray.prototype.size = function(d) {
    if (d != null) {
      return this._size[d];
    } else {
      return this._size;
    }
  };

  TypedArray.prototype.set_val = function(index, value) {
    index = this._get_index(index);
    if (this._data[index] !== value) {
      this._data[index] = value;
      return this._signal_change();
    }
  };

  TypedArray.prototype.nb_items = function() {
    var i, len, q, ref, tot;
    tot = this._size[0] || 0;
    ref = this._size.slice(1);
    for (q = 0, len = ref.length; q < len; q++) {
      i = ref[q];
      tot *= i;
    }
    return tot;
  };

  TypedArray.prototype.toString = function() {
    var i, j, l, len, m, o, q, ref, ref1, res, s, t, v;
    m = 1;
    res = "";
    l = (function() {
      var len, q, ref, results;
      ref = this._size;
      results = [];
      for (q = 0, len = ref.length; q < len; q++) {
        s = ref[q];
        o = m;
        m *= s;
        results.push(o);
      }
      return results;
    }).call(this);
    ref = this._data;
    for (i = q = 0, len = ref.length; q < len; i = ++q) {
      v = ref[i];
      res += v;
      for (j = t = ref1 = l.length - 1; ref1 <= 0 ? t <= 0 : t >= 0; j = ref1 <= 0 ? ++t : --t) {
        if (i % l[j] === l[j] - 1) {
          res += [" ", "\n", "\n\n"][j];
          break;
        }
      }
    }
    return res;
  };

  TypedArray.prototype.equals = function(obj) {
    var i, len, q, ref, v;
    if (obj instanceof TypedArray) {
      if (this._size.length !== obj._size.length) {
        return false;
      }
      ref = this._size;
      for (i = q = 0, len = ref.length; q < len; i = ++q) {
        v = ref[i];
        if (v !== obj._size[i]) {
          return false;
        }
      }
      return this._data === obj._data;
    }
    return this._data === obj;
  };

  TypedArray.prototype.get = function(index) {
    if (index != null) {
      return this._data[this._get_index(index)];
    } else {
      return this._data;
    }
  };

  TypedArray.prototype.resize = function(new_size) {
    var B, len, n, q, s, tot;
    tot = 1;
    for (q = 0, len = new_size.length; q < len; q++) {
      s = new_size[q];
      tot *= s;
    }
    B = this.base_type();
    n = new B(tot);
    n.set(this._data);
    this._data = n;
    this._size = new_size;
    return this._signal_change();
  };

  TypedArray.prototype._set = function(str) {
    var B;
    if (typeof str === "string") {
      this._set_state(str, {});
      return true;
    }
    if (this._data !== value || this._size.length !== 1 || this._size[0] !== value.length) {
      B = this.base_type();
      this._data = new B(value);
      this._size = [value.length];
      return true;
    }
    return false;
  };

  TypedArray.prototype._get_index = function(index) {
    var i, m, o, q, ref;
    if (index.length) {
      o = 0;
      m = 1;
      for (i = q = 0, ref = index.length; 0 <= ref ? q < ref : q > ref; i = 0 <= ref ? ++q : --q) {
        o += m * index[i];
        m *= this._size[i];
      }
      return o;
    }
    return index;
  };

  TypedArray.prototype._get_fs_data = function(out) {
    FileSystem.set_server_id_if_necessary(out, this);
    return out.mod += "C " + this._server_id + " " + (this._get_state()) + " ";
  };

  TypedArray.prototype._get_state = function() {
    var d, len, len1, q, ref, ref1, res, s, t;
    res = "";
    res += this._size.length;
    ref = this._size;
    for (q = 0, len = ref.length; q < len; q++) {
      s = ref[q];
      res += "," + s;
    }
    ref1 = this._data;
    for (t = 0, len1 = ref1.length; t < len1; t++) {
      d = ref1[t];
      res += "," + d;
    }
    return res;
  };

  TypedArray.prototype._set_state = function(str, map) {
    var B, l, n, q, ref, results, s, v;
    l = str.split(",");
    s = parseInt(l[0]);
    this._size = (function() {
      var q, ref, results;
      results = [];
      for (v = q = 0, ref = s; 0 <= ref ? q < ref : q > ref; v = 0 <= ref ? ++q : --q) {
        results.push(parseInt(l[v + 1]));
      }
      return results;
    })();
    B = this.base_type();
    n = this.nb_items();
    this._data = new B(n);
    results = [];
    for (v = q = 0, ref = n; 0 <= ref ? q < ref : q > ref; v = 0 <= ref ? ++q : --q) {
      results.push(this._data[v] = parseFloat(l[s + 1 + v]));
    }
    return results;
  };

  return TypedArray;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.TypedArray_Float32 = (function(superClass) {
  extend(TypedArray_Float32, superClass);

  function TypedArray_Float32(size, data) {
    if (size == null) {
      size = [];
    }
    TypedArray_Float32.__super__.constructor.call(this, size, data);
  }

  TypedArray_Float32.prototype.base_type = function() {
    return Float32Array;
  };

  TypedArray_Float32.prototype.deep_copy = function() {
    return new TypedArray_Float32(this._size, this._data);
  };

  return TypedArray_Float32;

})(TypedArray);

root = typeof _root_obj === "undefined" ? global : window;

root.TypedArray_Float64 = (function(superClass) {
  extend(TypedArray_Float64, superClass);

  function TypedArray_Float64(size, data) {
    if (size == null) {
      size = [];
    }
    TypedArray_Float64.__super__.constructor.call(this, size, data);
  }

  TypedArray_Float64.prototype.base_type = function() {
    return Float64Array;
  };

  TypedArray_Float64.prototype.deep_copy = function() {
    return new TypedArray_Float64(this._size, this._data);
  };

  return TypedArray_Float64;

})(TypedArray);

root = typeof _root_obj === "undefined" ? global : window;

root.TypedArray_Int32 = (function(superClass) {
  extend(TypedArray_Int32, superClass);

  function TypedArray_Int32(size, data) {
    if (size == null) {
      size = [];
    }
    TypedArray_Int32.__super__.constructor.call(this, size, data);
  }

  TypedArray_Int32.prototype.base_type = function() {
    return Int32Array;
  };

  TypedArray_Int32.prototype.deep_copy = function() {
    return new TypedArray_Int32(this._size, this._data);
  };

  return TypedArray_Int32;

})(TypedArray);

root = typeof _root_obj === "undefined" ? global : window;

root.FileSystem = (function() {
  var XMLHttpRequest_node;

  FileSystem._cur_tmp_server_id = 0;

  FileSystem._sig_server = true;

  FileSystem._disp = false;

  FileSystem._userid = "644";

  if (typeof global !== 'undefined') {
    XMLHttpRequest_node = require('xhr2');
    FileSystem._XMLHttpRequest = XMLHttpRequest_node;
  }

  FileSystem._objects_to_send = {};

  FileSystem._timer_send = void 0;

  FileSystem._timer_chan = void 0;

  FileSystem._nb_callbacks = 0;

  FileSystem._callbacks = {};

  FileSystem._nb_insts = 0;

  FileSystem._insts = {};

  FileSystem._files_to_upload = {};

  FileSystem._ptr_to_update = {};

  FileSystem._tmp_objects = {};

  FileSystem._objects = {};

  FileSystem._url = "127.0.0.1";

  FileSystem._port = "8888";

  FileSystem.url_com = "/sceen/_";

  FileSystem.url_upload = "/sceen/upload";

  if (typeof global !== 'undefined') {
    FileSystem.CONNECTOR_TYPE = "Node";
  } else {
    FileSystem.CONNECTOR_TYPE = "Browser";
  }

  function FileSystem() {
    this._data_to_send = "";
    this._session_num = -2;
    this._num_inst = FileSystem._nb_insts++;
    FileSystem._insts[this._num_inst] = this;
    if (FileSystem._userid != null) {
      this.send("U " + FileSystem._userid + " " + FileSystem._password + " ");
    }
    this.send("S " + this._num_inst + " ");
  }

  FileSystem.prototype.load = function(path, callback) {
    FileSystem._send_chan();
    this.send("L " + FileSystem._nb_callbacks + " " + (encodeURI(path)) + " ");
    FileSystem._callbacks[FileSystem._nb_callbacks] = callback;
    return FileSystem._nb_callbacks++;
  };

  FileSystem.prototype.load_or_make_dir = function(dir, callback) {
    return this.load(dir, (function(_this) {
      return function(res, err) {
        var lst, nir, oir, v;
        if (err) {
          if (dir === "/") {
            return callback(0, err);
          } else {
            lst = (function() {
              var len, q, ref, results;
              ref = dir.split('/');
              results = [];
              for (q = 0, len = ref.length; q < len; q++) {
                v = ref[q];
                if (v.length) {
                  results.push(v);
                }
              }
              return results;
            })();
            nir = lst.pop();
            oir = "/" + lst.join("/");
            return _this.load_or_make_dir(oir, function(n_res, n_err) {
              var n_dir;
              if (n_err) {
                return callback(0, n_err);
              } else {
                n_dir = new Directory;
                n_res.add_file(nir, n_dir);
                return callback(n_dir, n_err);
              }
            });
          }
        } else {
          return callback(res, err);
        }
      };
    })(this));
  };

  FileSystem.prototype.load_ptr = function(ptr, callback) {
    FileSystem._send_chan();
    this.send("l " + FileSystem._nb_callbacks + " " + ptr + " ");
    FileSystem._callbacks[FileSystem._nb_callbacks] = callback;
    return FileSystem._nb_callbacks++;
  };

  FileSystem.prototype.send = function(data) {
    this._data_to_send += data;
    if (FileSystem._timer_send == null) {
      return FileSystem._timer_send = setTimeout(FileSystem._timeout_send_func, 1);
    }
  };

  FileSystem.prototype.make_channel = function() {
    var path, xhr_object;
    path = "";
    if (FileSystem.CONNECTOR_TYPE === "Browser") {
      path = FileSystem.url_com + ("?s=" + this._session_num);
    } else if (FileSystem.CONNECTOR_TYPE === "Node") {
      path = "http://" + FileSystem._url + ":" + FileSystem._port + FileSystem.url_com + ("?s=" + this._session_num);
    }
    xhr_object = FileSystem._my_xml_http_request();
    xhr_object.open('GET', path, true);
    xhr_object.onreadystatechange = function() {
      var _w;
      if (this.readyState === 4 && this.status === 200) {
        if (FileSystem._disp) {
          console.log("chan ->", this.responseText);
        }
        _w = function(sid, obj) {
          if ((sid != null) && (obj != null)) {
            obj._server_id = sid;
            return FileSystem._objects[sid] = obj;
          }
        };
        FileSystem._sig_server = false;
        eval(this.responseText);
        return FileSystem._sig_server = true;
      }
    };
    return xhr_object.send();
  };

  FileSystem.get_inst = function() {
    var i, k, ref;
    ref = FileSystem._insts;
    for (k in ref) {
      i = ref[k];
      return i;
    }
    return new FileSystem;
  };

  FileSystem.set_server_id_if_necessary = function(out, obj) {
    var ncl;
    if (obj._server_id == null) {
      obj._server_id = FileSystem._get_new_tmp_server_id();
      FileSystem._tmp_objects[obj._server_id] = obj;
      ncl = ModelProcessManager.get_object_class(obj);
      if (obj._underlying_fs_type != null) {
        out.mod += "T " + obj._server_id + " " + ncl + " ";
        ncl = obj._underlying_fs_type();
      }
      out.cre += "N " + obj._server_id + " " + ncl + " ";
      return obj._get_fs_data(out);
    }
  };

  FileSystem.signal_change = function(m) {
    if (FileSystem._sig_server) {
      FileSystem._objects_to_send[m.model_id] = m;
      if (FileSystem._timer_chan != null) {
        clearTimeout(FileSystem._timer_chan);
      }
      return FileSystem._timer_chan = setTimeout(FileSystem._timeout_chan_func, 250);
    }
  };

  FileSystem._tmp_id_to_real = function(tmp_id, res) {
    var fs, path, ptr, tmp, xhr_object;
    tmp = FileSystem._tmp_objects[tmp_id];
    if (tmp == null) {
      console.log(tmp_id);
    }
    FileSystem._objects[res] = tmp;
    tmp._server_id = res;
    delete FileSystem._tmp_objects[tmp_id];
    ptr = FileSystem._ptr_to_update[tmp_id];
    if (ptr != null) {
      delete FileSystem._ptr_to_update[tmp_id];
      ptr.data.value = res;
    }
    if ((FileSystem._files_to_upload[tmp_id] != null) && (tmp.file != null)) {
      delete FileSystem._files_to_upload[tmp_id];
      fs = FileSystem.get_inst();
      path = "";
      if (FileSystem.CONNECTOR_TYPE === "Browser") {
        path = FileSystem.url_com + ("?s=" + fs._session_num + "&p=" + tmp._server_id);
      } else if (FileSystem.CONNECTOR_TYPE === "Node") {
        path = "http://" + FileSystem._url + ":" + FileSystem._port + FileSystem.url_com + ("?s=" + fs._session_num + "&p=" + tmp._server_id);
      }
      xhr_object = FileSystem._my_xml_http_request();
      xhr_object.open('PUT', path, true);
      xhr_object.onreadystatechange = function() {
        var _w;
        if (this.readyState === 4 && this.status === 200) {
          _w = function(sid, obj) {
            if ((sid != null) && (obj != null)) {
              obj._server_id = sid;
              return FileSystem._objects[sid] = obj;
            }
          };
          return eval(this.responseText);
        }
      };
      xhr_object.send(tmp.file);
      return delete tmp.file;
    }
  };

  FileSystem._get_new_tmp_server_id = function() {
    FileSystem._cur_tmp_server_id++;
    if (FileSystem._cur_tmp_server_id % 4 === 0) {
      FileSystem._cur_tmp_server_id++;
    }
    return FileSystem._cur_tmp_server_id;
  };

  FileSystem._send_chan = function() {
    var f, k, out, ref, results;
    out = FileSystem._get_chan_data();
    ref = FileSystem._insts;
    results = [];
    for (k in ref) {
      f = ref[k];
      results.push(f.send(out));
    }
    return results;
  };

  FileSystem._timeout_chan_func = function() {
    FileSystem._send_chan();
    return delete FileSystem._timer_chan;
  };

  FileSystem._get_chan_data = function() {
    var model, n, out, ref;
    out = {
      cre: "",
      mod: ""
    };
    ref = FileSystem._objects_to_send;
    for (n in ref) {
      model = ref[n];
      model._get_fs_data(out);
    }
    FileSystem._objects_to_send = {};
    return out.cre + out.mod;
  };

  FileSystem._timeout_send_func = function() {
    var f, k, out, path, ref, ref1, xhr_object;
    out = FileSystem._get_chan_data();
    ref = FileSystem._insts;
    for (k in ref) {
      f = ref[k];
      f._data_to_send += out;
    }
    ref1 = FileSystem._insts;
    for (k in ref1) {
      f = ref1[k];
      if (!f._data_to_send.length) {
        continue;
      }
      if (f._session_num === -1) {
        continue;
      }
      if (f._session_num === -2) {
        f._session_num = -1;
      } else {
        f._data_to_send = ("s " + f._session_num + " ") + f._data_to_send;
      }
      path = "";
      if (FileSystem.CONNECTOR_TYPE === "Browser") {
        path = FileSystem.url_com;
      } else if (FileSystem.CONNECTOR_TYPE === "Node") {
        path = "http://" + FileSystem._url + ":" + FileSystem._port + FileSystem.url_com;
      }
      xhr_object = FileSystem._my_xml_http_request();
      xhr_object.open('POST', path, true);
      xhr_object.onreadystatechange = function() {
        var _c, _w, c, len, q, results;
        if (this.readyState === 4 && this.status === 200) {
          if (FileSystem._disp) {
            console.log("resp ->", this.responseText);
          }
          _c = [];
          _w = function(sid, obj) {
            if ((sid != null) && (obj != null)) {
              obj._server_id = sid;
              return FileSystem._objects[sid] = obj;
            }
          };
          FileSystem._sig_server = false;
          eval(this.responseText);
          FileSystem._sig_server = true;
          results = [];
          for (q = 0, len = _c.length; q < len; q++) {
            c = _c[q];
            results.push(FileSystem._callbacks[c[0]](FileSystem._objects[c[1]], c[2]));
          }
          return results;
        }
      };
      if (FileSystem._disp) {
        console.log("sent ->", f._data_to_send + "E ");
      }
      xhr_object.setRequestHeader('Content-Type', 'text/plain');
      xhr_object.send(f._data_to_send + "E ");
      f._data_to_send = "";
    }
    FileSystem._objects_to_send = {};
    return delete FileSystem._timer_send;
  };

  FileSystem._my_xml_http_request = function() {
    if (FileSystem.CONNECTOR_TYPE === "Browser") {
      if (window.XMLHttpRequest) {
        return new XMLHttpRequest;
      }
      if (window.ActiveXObject) {
        return new ActiveXObject('Microsoft.XMLHTTP');
      }
      return alert('Your browser does not seem to support XMLHTTPRequest objects...');
    } else if (FileSystem.CONNECTOR_TYPE === "Node") {
      return new FileSystem._XMLHttpRequest();
    } else {
      return console.log("you must define CONNECTOR_TYPE");
    }
  };

  return FileSystem;

})();

root = typeof _root_obj === "undefined" ? global : window;

root.Directory = (function(superClass) {
  extend(Directory, superClass);

  function Directory() {
    Directory.__super__.constructor.call(this);
  }

  Directory.prototype.base_type = function() {
    return File;
  };

  Directory.prototype.find = function(name) {
    var f, len, q;
    for (q = 0, len = this.length; q < len; q++) {
      f = this[q];
      if (f.name.equals(name)) {
        return f;
      }
    }
    return void 0;
  };

  Directory.prototype.load = function(name, callback) {
    var f;
    f = this.find(name);
    if (f) {
      return f.load(callback);
    } else {
      return callback(void 0, "file does not exist");
    }
  };

  Directory.prototype.has = function(name) {
    var f, len, q;
    for (q = 0, len = this.length; q < len; q++) {
      f = this[q];
      if (f.name.equals(name)) {
        return true;
      }
    }
    return false;
  };

  Directory.prototype.add_file = function(name, obj, params) {
    var o, res;
    if (params == null) {
      params = {};
    }
    o = this.find(name);
    if (o != null) {
      return o;
    }
    res = new File(name, obj, params);
    this.push(res);
    return res;
  };

  Directory.prototype.add_tiff_file = function(name, obj, tiff_obj, params) {
    var o, res;
    if (params == null) {
      params = {};
    }
    o = this.find(name);
    if (o != null) {
      return o;
    }
    res = new TiffFile(name, obj, tiff_obj, params);
    this.push(res);
    return res;
  };

  Directory.prototype.force_add_file = function(name, obj, params) {
    var find_name, name_file, num, o, res;
    if (params == null) {
      params = {};
    }
    num = this.length;
    find_name = true;
    name_file = name + "_" + num;
    while (find_name) {
      name_file = name + "_" + num;
      o = this.find(name_file);
      if (o != null) {
        num += 1;
      } else {
        break;
      }
    }
    console.log(name_file);
    res = new File(name_file, obj, params);
    this.push(res);
    return res;
  };

  Directory.prototype.get_file_info = function(info) {
    return info.icon = "folder";
  };

  return Directory;

})(Lst);

root = typeof _root_obj === "undefined" ? global : window;

root.File = (function(superClass) {
  extend(File, superClass);

  function File(name, ptr_or_model, info) {
    var cp_info, key, val;
    if (name == null) {
      name = "";
    }
    if (ptr_or_model == null) {
      ptr_or_model = 0;
    }
    if (info == null) {
      info = {};
    }
    File.__super__.constructor.call(this);
    cp_info = {};
    for (key in info) {
      val = info[key];
      cp_info[key] = val;
    }
    if (ptr_or_model instanceof Model) {
      if (cp_info.model_type == null) {
        cp_info.model_type = ModelProcessManager.get_object_class(ptr_or_model);
      }
      if (typeof ptr_or_model.get_file_info === "function") {
        ptr_or_model.get_file_info(cp_info);
      }
    }
    this.add_attr({
      name: name,
      admins: new Lst,
      users: new Lst,
      _created_at: new Date(),
      _ptr: new Ptr(ptr_or_model),
      _info: cp_info
    });
  }

  File.prototype.load = function(callback) {
    return this._ptr.load(callback);
  };

  return File;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.Path = (function(superClass) {
  extend(Path, superClass);

  function Path(file1) {
    var size;
    this.file = file1;
    Path.__super__.constructor.call(this);
    size = this.file != null ? this.file.fileSize != null ? this.file.fileSize : this.file.size : 0;
    this.add_attr({
      remaining: size,
      to_upload: size
    });
  }

  Path.prototype.get_file_info = function(info) {
    info.remaining = this.remaining;
    return info.to_upload = this.to_upload;
  };

  Path.prototype._get_fs_data = function(out) {
    Path.__super__._get_fs_data.call(this, out);
    if ((this.file != null) && this._server_id & 3) {
      return FileSystem._files_to_upload[this._server_id] = this;
    }
  };

  return Path;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.Ptr = (function(superClass) {
  extend(Ptr, superClass);

  function Ptr(model) {
    Ptr.__super__.constructor.call(this);
    this.data = {};
    this._set(model);
  }

  Ptr.prototype.load = function(callback) {
    var ref;
    if (this.data.model != null) {
      return callback(this.data.model, false);
    } else {
      return (ref = FileSystem.get_inst()) != null ? ref.load_ptr(this.data.value, callback) : void 0;
    }
  };

  Ptr.prototype._get_fs_data = function(out) {
    FileSystem.set_server_id_if_necessary(out, this);
    if (this.data.model != null) {
      FileSystem.set_server_id_if_necessary(out, this.data.model);
      out.mod += "C " + this._server_id + " " + this.data.model._server_id + " ";
      this.data.value = this.data.model._server_id;
      if (this.data.model._server_id & 3) {
        return FileSystem._ptr_to_update[this.data.model._server_id] = this;
      }
    } else {
      return out.mod += "C " + this._server_id + " " + this.data.value + " ";
    }
  };

  Ptr.prototype._set = function(model) {
    var res;
    if (typeof model === "number") {
      res = this.data.value !== model;
      this.data = {
        value: model
      };
      return res;
    }
    if (model instanceof Model) {
      res = this.data.value !== model._server_id;
      this.data = {
        model: model,
        value: model._server_id
      };
      return res;
    }
    return false;
  };

  Ptr.prototype._get_state = function() {
    return this._data;
  };

  Ptr.prototype._set_state = function(str, map) {
    return this.set(str);
  };

  return Ptr;

})(Model);

root = typeof _root_obj === "undefined" ? global : window;

root.TiffFile = (function(superClass) {
  extend(TiffFile, superClass);

  function TiffFile(name, ptr_or_model, ptr_tiff, info) {
    if (name == null) {
      name = "";
    }
    if (ptr_or_model == null) {
      ptr_or_model = 0;
    }
    if (ptr_tiff == null) {
      ptr_tiff = 0;
    }
    if (info == null) {
      info = {};
    }
    TiffFile.__super__.constructor.call(this, name, ptr_or_model, info);
    this.add_attr({
      _ptr_tiff: new Ptr(ptr_tiff),
      _has_been_converted: 0
    });
  }

  TiffFile.prototype.load_tiff = function(callback) {
    return this._ptr_tiff.load(callback);
  };

  return TiffFile;

})(File);

root = typeof _root_obj === "undefined" ? global : window;

root.Process = (function() {
  function Process(m, onchange_construction) {
    var i, len, q;
    if (onchange_construction == null) {
      onchange_construction = true;
    }
    this.process_id = ModelProcessManager._cur_process_id;
    ModelProcessManager._cur_process_id += 1;
    this._models = [];
    if (m instanceof Model) {
      m.bind(this, onchange_construction);
    } else if (m.length != null) {
      for (q = 0, len = m.length; q < len; q++) {
        i = m[q];
        i.bind(this, onchange_construction);
      }
    } else if (m != null) {
      console.error("Process constructor doesn't know what to do with", m);
    }
  }

  Process.prototype.destructor = function() {
    var i, len, m, q, ref, results;
    ref = this._models;
    results = [];
    for (q = 0, len = ref.length; q < len; q++) {
      m = ref[q];
      i = m._processes.indexOf(this);
      if (i >= 0) {
        results.push(m._processes.splice(i, 1));
      } else {
        results.push(void 0);
      }
    }
    return results;
  };

  Process.prototype.onchange = function() {};

  return Process;

})();

root.bind = (function(_this) {
  return function(m, f) {
    var i, len, q, results;
    if (m instanceof Model) {
      return m.bind(f);
    } else {
      results = [];
      for (q = 0, len = m.length; q < len; q++) {
        i = m[q];
        results.push(i.bind(f));
      }
      return results;
    }
  };
})(this);

root = typeof _root_obj === "undefined" ? global : window;

root.BindProcess = (function(superClass) {
  extend(BindProcess, superClass);

  function BindProcess(model, onchange_construction, f1) {
    this.f = f1;
    BindProcess.__super__.constructor.call(this, model, onchange_construction);
  }

  BindProcess.prototype.onchange = function() {
    return this.f();
  };

  return BindProcess;

})(Process);
